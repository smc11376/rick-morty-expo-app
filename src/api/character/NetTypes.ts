export type CharacterNetType = {
  id: number;
  name: string;
  status: string;
  species: string;
  type: string;
  gender: string;
  origin: OriginOrLocationNetType;
  location: OriginOrLocationNetType;
  image: string;
  episode?: string[] | null;
  url: string;
  created: string;
};

export type OriginOrLocationNetType = {
  name: string;
  url: string;
};

export type GetAllCharactersNetType = {
  info: {
    count: number;
    pages: number;
    next: string | null;
    prev: string | null;
  };
  results: CharacterNetType[];
};
