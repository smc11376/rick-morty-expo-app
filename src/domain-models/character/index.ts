export type ItemWithUrl = {
  name: string;
  url: string;
};

export type ShortCharacter = {
  id: number;
  name: string;
  status: string;
  image: string;
};

export type Character = ShortCharacter & {
  species: string;
  type: string;
  gender: string;
  origin: ItemWithUrl;
  location: ItemWithUrl;
  numberOfEpisodes: number;
};
