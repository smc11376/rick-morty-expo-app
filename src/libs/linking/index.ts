import * as Linking from "expo-linking";

export function openUrl(url?: string) {
  if (!url) return;
  Linking.canOpenURL(url).then(canOpen => {
    if (canOpen) Linking.openURL(url);
  });
}
