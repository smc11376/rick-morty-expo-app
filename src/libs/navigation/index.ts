import { router } from "expo-router";
import { CHARACTER_DETAIL_PATH, HOME_PATH } from "@/libs/navigation/pathRoutes";

function goToPath(path: string, params?: any) {
  router.navigate(path);
  router.setParams(params);
}

export function goBack() {
  router.back();
}

export function goToHome() {
  goToPath(HOME_PATH);
}

export function goToCharacterDetail(id: number, characterName: string) {
  goToPath(CHARACTER_DETAIL_PATH.replace("[id]", id.toString()), { characterName });
}
