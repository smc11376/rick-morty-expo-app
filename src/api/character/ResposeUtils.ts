import { CharacterNetType, GetAllCharactersNetType } from "@/api/character/NetTypes";
import { Character, ShortCharacter } from "@/domain-models/character";

export function transformGetAllCharactersResponse(response?: GetAllCharactersNetType): ShortCharacter[] {
  if (!response) return [];

  return response.results.map(item => ({
    id: item.id,
    name: item.name,
    status: item.status,
    image: item.image,
  }));
}

export function transformGetCharacterByIdResponse(response?: CharacterNetType): Character | undefined {
  if (!response) return undefined;

  return {
    id: response.id,
    name: response.name,
    status: response.status,
    species: response.species,
    type: response.type,
    gender: response.gender,
    origin: response.origin,
    location: response.location,
    image: response.image,
    numberOfEpisodes: response.episode?.length || 0,
  };
}
