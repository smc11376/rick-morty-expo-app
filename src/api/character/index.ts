import { useGetApiService } from "@/libs/network";
import { CharacterNetType, GetAllCharactersNetType } from "@/api/character/NetTypes";
import { transformGetAllCharactersResponse, transformGetCharacterByIdResponse } from "@/api/character/ResposeUtils";
import { Character, ShortCharacter } from "@/domain-models/character";
import { useEffect, useMemo, useState } from "react";

type UseAllCharactersType = {
  characters: ShortCharacter[];
  nextPage: string | null | undefined;
  isLoading: boolean;
};

type UseCharacterType = {
  character: Character | undefined;
  isLoading: boolean;
};

export function useGetAllCharacters(newUrl: string | null | undefined): UseAllCharactersType {
  const [characters, setCharacters] = useState<ShortCharacter[]>([]);
  const { data, isLoading, isFetching, isRefetching } = useGetApiService<GetAllCharactersNetType>(
    newUrl || "/character",
  );
  const newCharacters = useMemo(() => transformGetAllCharactersResponse(data), [data?.results]);

  useEffect(() => {
    if (!isLoading && !isFetching && !isRefetching) {
      setCharacters(currentCharacters => {
        return [...new Set([...currentCharacters, ...newCharacters])];
      });
    }
  }, [newCharacters]);

  return {
    characters,
    isLoading: isLoading || isFetching || isRefetching,
    nextPage: data?.info.next,
  };
}

export function useGetCharacterById(id?: string): UseCharacterType {
  if (!id) throw new Error("Oops! There is no character to find!");

  const { data, isLoading, isFetching, isRefetching } = useGetApiService<CharacterNetType>(`/character/${id}`);
  const character = transformGetCharacterByIdResponse(data);
  return {
    character,
    isLoading: isLoading || isFetching || isRefetching,
  };
}
