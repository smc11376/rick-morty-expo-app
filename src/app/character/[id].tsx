import { ScrollView, StyleSheet, View } from "react-native";
import { useGlobalSearchParams, useNavigation } from "expo-router";
import { useGetCharacterById } from "@/api/character";
import { useEffect } from "react";
import Loader from "@/components/loader";
import Row from "@/app/character/components/row";
import InfoBlock from "@/app/character/components/info-block";
import CharacterImage from "@/app/character/components/character-image";

export default function CharacterDetailScreen() {
  const params = useGlobalSearchParams();
  const { character, isLoading } = useGetCharacterById(params.id as string);
  const navigation = useNavigation();

  useEffect(() => {
    navigation.setOptions({ title: params.characterName });
  }, [params.characterName]);

  return isLoading ? (
    <View style={styles.container}>
      <Loader />
    </View>
  ) : (
    <ScrollView style={styles.container} contentContainerStyle={styles.contentContainerStyle}>
      <CharacterImage uri={character?.image} />
      <Row>
        <InfoBlock label="Gender: " data={character?.gender} />
        <InfoBlock label="Status: " data={character?.status} />
      </Row>
      <Row>
        <InfoBlock label="Species: " data={character?.species} />
        {character?.type !== "" && <InfoBlock label="Type: " data={character?.type} />}
      </Row>
      <InfoBlock label="Number of episodes: " data={character?.numberOfEpisodes.toString()} />
      <InfoBlock label="Origin: " data={character?.origin.name} urlToOpen={character?.origin.url} />
      <InfoBlock label="Current location: " data={character?.location.name} urlToOpen={character?.location.url} />
    </ScrollView>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: "rgb(60, 62, 68)",
  },
  contentContainerStyle: {
    flexGrow: 1,
  },
  image: {
    width: "100%",
    aspectRatio: 1,
  },
});
