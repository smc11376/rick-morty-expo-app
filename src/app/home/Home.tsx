import { FlatList, StyleSheet, View } from "react-native";
import { useGetAllCharacters } from "@/api/character";
import { useCallback, useState } from "react";
import CharacterCard from "@/components/character-card";
import { ShortCharacter } from "@/domain-models/character";
import { goToCharacterDetail } from "@/libs/navigation";
import HomeHeader from "@/app/home/components/HomeHeader";
import Loader from "@/components/loader";

export default function Home() {
  const [currentPage, setCurrentPage] = useState<string | null | undefined>(null);
  const { characters, isLoading, nextPage } = useGetAllCharacters(currentPage);

  const renderCharacterCard = useCallback(
    (item: ShortCharacter) => (
      <CharacterCard character={item} onPressCharacter={() => goToCharacterDetail(item.id, item.name)} />
    ),
    [characters],
  );

  function handleLoadMore() {
    !!nextPage && setCurrentPage(nextPage);
  }

  return (
    <View style={styles.container}>
      <HomeHeader />
      <FlatList
        data={characters}
        renderItem={({ item }) => renderCharacterCard(item)}
        keyExtractor={item => item.id.toString()}
        onEndReached={handleLoadMore}
        onEndReachedThreshold={0.3}
        ListFooterComponent={() => (isLoading ? <Loader /> : null)}
      />
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: "rgb(39, 43, 51)",
  },
});
