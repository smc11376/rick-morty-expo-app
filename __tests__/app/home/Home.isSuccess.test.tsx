import { fireEvent, render, screen, waitFor } from "@testing-library/react-native";
import Home from "@/app/home/Home";
import { ShortCharacter } from "@/domain-models/character";
import * as Navigation from "@/libs/navigation";

const characterMock: ShortCharacter = {
  id: 1,
  name: "Testing Name",
  status: "Alive",
  image: "https://rickandmortyapi.com/api/character/avatar/400.jpeg",
};

jest.mock("@/api/character", () => ({
  useGetAllCharacters: () => ({
    characters: [characterMock],
    isLoading: false,
    nextPage: null,
  }),
}));

test("Home renders isSuccess", () => {
  render(<Home />);

  expect(screen.toJSON()).toMatchSnapshot();
});

test("Home actions", async () => {
  const goToCharacterDetailMock = jest.spyOn(Navigation, "goToCharacterDetail");
  render(<Home />);

  fireEvent.press(screen.getAllByTestId("character-card-wrapper")[0]);
  await waitFor(() => expect(goToCharacterDetailMock).toHaveBeenCalled());
});
