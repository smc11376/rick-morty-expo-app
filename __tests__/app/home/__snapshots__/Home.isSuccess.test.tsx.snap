// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home renders isSuccess 1`] = `
<View
  style={
    {
      "backgroundColor": "rgb(39, 43, 51)",
      "flex": 1,
    }
  }
>
  <View
    style={
      [
        {
          "alignItems": "center",
          "backgroundColor": "white",
          "height": 160,
          "justifyContent": "center",
          "width": "100%",
        },
        {
          "paddingTop": 20,
        },
      ]
    }
  >
    <Text
      style={
        {
          "color": "black",
          "fontSize": 24,
          "fontWeight": "bold",
        }
      }
    >
      The Rick and Morty API
    </Text>
  </View>
  <RCTScrollView
    ListFooterComponent={[Function]}
    data={
      [
        {
          "id": 1,
          "image": "https://rickandmortyapi.com/api/character/avatar/400.jpeg",
          "name": "Testing Name",
          "status": "Alive",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onEndReached={[Function]}
    onEndReachedThreshold={0.3}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={0.0001}
    stickyHeaderIndices={[]}
    viewabilityConfigCallbackPairs={[]}
  >
    <View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
              "paddingHorizontal": 24,
              "paddingVertical": 8,
            }
          }
          testID="character-card-wrapper"
        >
          <View
            style={
              {
                "alignItems": "center",
                "backgroundColor": "rgb(60, 62, 68)",
                "borderRadius": 8,
                "flexDirection": "row",
                "minHeight": 80,
                "width": "100%",
              }
            }
          >
            <Image
              resizeMode="contain"
              source={
                {
                  "uri": "https://rickandmortyapi.com/api/character/avatar/400.jpeg",
                }
              }
              style={
                {
                  "borderBottomLeftRadius": 8,
                  "borderTopLeftRadius": 8,
                  "height": 80,
                  "width": 80,
                }
              }
            />
            <View
              style={
                {
                  "flex": 1,
                  "padding": 8,
                }
              }
            >
              <Text
                style={
                  {
                    "color": "white",
                    "flex": 1,
                    "fontSize": 16,
                    "fontWeight": "bold",
                  }
                }
              >
                Testing Name
              </Text>
              <Text
                style={
                  {
                    "color": "white",
                    "fontSize": 12,
                  }
                }
              >
                Status: 
                <Text
                  style={
                    {
                      "color": "green",
                    }
                  }
                >
                  Alive
                </Text>
              </Text>
            </View>
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
      />
    </View>
  </RCTScrollView>
</View>
`;
